helm-charts/orchestrator/templates/network-policies.yaml

```diff
@@ -12,7 +12,7 @@ spec:
       - namespaceSelector:
           matchLabels:
             # Allow RHDH namespace to communicate with workflows and sonataflow services
-            kubernetes.io/metadata.name: {{ .Values.rhdhOperator.subscription.targetNamespace }}
+            kubernetes.io/metadata.name: {{- $namespace := or (.Values.rhdhOperator.namespace) (and (.Values.rhdhOperator.subscription) (.Values.rhdhOperator.subscription.targetNamespace)) "rhdh-operator" }}{{ $namespace }}
```

Comment:
Just for the sake of readability, can we go with something like:

```
{{- /* Set $namespace to the defined namespace value or a fallback */ -}}
{{- $namespace := or 
    .Values.rhdhOperator.namespace                 # Use rhdhOperator.namespace if defined
    (and                                           # Otherwise, if rhdhOperator.subscription is defined,
        .Values.rhdhOperator.subscription          # check if targetNamespace is also set,
        .Values.rhdhOperator.subscription.targetNamespace
    )
    "rhdh-operator"                                # Fallback to "rhdh-operator" if none of the above are set
-}}
{{ $namespace }}
```

Reply:
I'm confused about the use case here: are we expecting the user to have pre-installed the rhdh operator prior to deployment of our CR? if that's the case can we check if the subscription exists in the given `existingRHDHNamespace` value and if so, then we proceed to ignore our subscription values.
Or perhaps it is possible for the user to install RHDH after our CR is deployed? In that case there's no need to check for subscription and what is done here is sufficient. Thoughs?

Reply:
There's a small problem having this new variable around: the rhdh operator chart uses the `subscription.targetNamespace` to check for the contents of the [secret](https://github.com/parodos-dev/orchestrator-helm-operator/blob/d480506d9549d219edb3201f7be48a9f06318841/helm-charts/orchestrator/templates/rhdh-operator.yaml#L28). If we allow the user to reference a different location for the namespace, we should also take this into account when checking for the contents of the secret.

I suggest to encapsulate the logic to extract the location of the namespace in a function in `_helpers.tpl` that can be reused across the charts and point to the correct location of rhdh based on the logic defined in this line.

Reply:
@jenniferubah what do you think to simplify the expression like this?
```
kubernetes.io/metadata.name: {{ or (.Values.rhdhOperator.existingRHDHNamespace) (and (.Values.rhdhOperator.subscription) (.Values.rhdhOperator.subscription.targetNamespace)) }}
```
Also, keep in mind that the dash in `{{-` will remove any space between the output of the expression and the character before that. To be yaml compliant, we want to have a space between the colon `:` and the value:
```
kubernetes.io/metadata.name: rhdh-operator
```

To test it, add this expression at the bottom of the NOTES.txt and then run this command:
```
helm install --dry-run=server orchestrator helm-charts/orchestrator
```

Reply:
> I'm confused about the use case here: are we expecting the user to have pre-installed the rhdh operator prior to deployment of our CR? if that's the case can we check if the subscription exists in the given `existingRHDHNamespace` value and if so, then we proceed to ignore our subscription values. Or perhaps it is possible for the user to install RHDH after our CR is deployed? In that case there's no need to check for subscription and what is done here is sufficient. Thoughs?

We are attempting to handle two use cases for the network policy aspect;
- For the case of the pre existing RHDH instance (installed prior to our CR), we want to use the value of `existingRHDHNamespace`
- For the case of new RHDH instance (installed by our CR), we want to use our subscription values. That's why we check for it.

Reply:
> There's a small problem having this new variable around: the rhdh operator chart uses the `subscription.targetNamespace` to check for the contents of the [secret](https://github.com/parodos-dev/orchestrator-helm-operator/blob/d480506d9549d219edb3201f7be48a9f06318841/helm-charts/orchestrator/templates/rhdh-operator.yaml#L28). If we allow the user to reference a different location for the namespace, we should also take this into account when checking for the contents of the secret.
> 
> I suggest to encapsulate the logic to extract the location of the namespace in a function in `_helpers.tpl` that can be reused across the charts and point to the correct location of rhdh based on the logic defined in this line.

Currently, this value is only used within the network policy manifest. And the goal is to allow the user input namespace only for an existing RHDH instance. The value of `subscription.targetNamespace` will still be used to check for the secrets when creating the RHDH Operator process.

Reply:
@masayag I had a conversation with @jenniferubah to discuss this PR and I ended up proposing this:
* `rhdhOperator.existingRHDHNamespace` to be moved to a new section `networkPolicy.rhdhNamespace`. The motivation is that the scope of the requirement is for `networkPolicy` to determine which namespace to use for rhdh. On its own, the `rhdhOperator` section makes no use of the new field, so it feels more approppiate that the value is captured in a new section defined as `networkPolicy`. So instead of:

```
rhdhOperator:
  enabled: false
  existingRHDHNamespace: my-rhdh-operator
```

it would be:

```
rhdhOperator:
  enabled: false

networkPolicy:
  rhdhNamespace: my-rhdh-operator
```

* Add a conditional to the logic to print the namespace to check if the `rhdhOperator.enabled` is false (=> use the value in `networkPolicy.rhdhNamespace`) or else use the value in `rhdhOperator.subscription.targetNamespace`). If the value of the `rhdhNamespace` or `targetNamespace` is empty in either case, fail the deployment. Something like this:

```
{{- $namespace:= "" }}
{{- if (and (not .Values.rhdhOperator.enabled) .Values.networkPolicy.rhdhNamespace) }}
  {{- $namespace = .Values.networkPolicy.rhdhNamespace }}
{{- else if (and (.Values.rhdhOperator.enabled) ( and .Values.rhdhOperator.subscription .Values.rhdhOperator.subscription.targetNamespace ) )}}
  {{- $namespace = .Values.rhdhOperator.subscription.targetNamespace }}
{{- else }}
   {{- fail "No RHDH namespace defined" }}
{{- end }}
kubernetes.io/metadata.name: {{- $namespace }}
```

What do you think?

Reply:
I like most it, but would like to take this one step further and split the value of the RHDH instance installed by our CR to its own property.
why? mainly since `rhdhOperator.subscription.targetNamespace` refers to the subscription namespace but shouldn't represent the target namespace of the RHDH instance created by the operator.
This might be a bit out-of-scope of this PR, but I'd be happy to include it as part of it instead of adding more logic on top of the current `rhdhOperator.subscription.targetNamespace` property.

The last thing I'd like to confirm is that we do not require to use the new `rhdhNamespace` outside of the `networkPolicy` section, otherwise, we'll have to introduce it differently. If there isn't any use for it outside of that section, than I'm almost okay with it, but still have a slight concern that we expose the implementation detail by using the `networkPolicy` section which perhaps the user shouldn't be aware of. Overall, the network policy will be created in the sonataflow-infra namespace.

Reply:
> I like most it, but would like to take this one step further and split the value of the RHDH instance installed by our CR to its own property.
why? mainly since rhdhOperator.subscription.targetNamespace refers to the subscription namespace but shouldn't represent the target namespace of the RHDH instance created by the operator.
>This might be a bit out-of-scope of this PR, but I'd be happy to include it as part of it instead of adding more logic on top of the current rhdhOperator.subscription.targetNamespace property.

We're adding this logic condition to the `networkPolicy` manifest only, there's no change to the other manifests that refer to this field. I'm not sure I'm following you about splitting the value of the RHDH instance. Maybe an example might help?

>The last thing I'd like to confirm is that we do not require to use the new rhdhNamespace outside of the networkPolicy section, otherwise, we'll have to introduce it differently. If there isn't any use for it outside of that section, than I'm almost okay with it, but still have a slight concern that we expose the implementation detail by using the networkPolicy section which perhaps the user shouldn't be aware of. Overall, the network policy will be created in the sonataflow-infra namespace.

Makes sense. With helm we have a few tricks in our sleeve to identify where RHDH is installed, but still that would not be accurate to assume that the one we find is the one the user wants to configure. The cluster might have multiple instances and we would not know which is the one we're targeting, hence the need to specify.

Ultimately we can always remove this field from the CRD and make it dependent on having the CR installing RHDH or not. If the CR has `rhdhOperator.enabled=false` then the user is responsible of configuring the `networkPolicy` on their own (documentation would be needed to provide a sample manifest for the neworkPolicy).

Reply:
> We're adding this logic condition to the networkPolicy manifest only, there's no change to the other manifests that refer to this field. I'm not sure I'm following you about splitting the value of the RHDH instance. Maybe an example might help?

I'd suggest to change https://github.com/parodos-dev/orchestrator-helm-operator/blob/main/helm-charts/orchestrator/templates/rhdh-operator.yaml
```
{{- define "backstage-manifest" }}
---
apiVersion: rhdh.redhat.com/v1alpha1
kind: Backstage
metadata:
  name: backstage
  namespace: {{ .Values.rhdhOperator.subscription.targetNamespace }}
  annotations:
    "meta.helm.sh/release-name": {{ .Release.Name}}
```

to 
```
{{- define "backstage-manifest" }}
---
apiVersion: rhdh.redhat.com/v1alpha1
kind: Backstage
metadata:
  name: backstage
  namespace: {{ .Values.rhdh.namespace }}
  annotations:
    "meta.helm.sh/release-name": {{ .Release.Name}}
```

meaning, replacement of `.Values.rhdhOperator.subscription.targetNamespace` with `.Values.rhdh.namespace`. We can propose a better name and/or path for it, but I'd like it not to be under the `rhdhOperator` section, therefore added the `rhdh` section. Then it makes sense to refer to it as a represent of the existing or the new RHDH instance.

e.g. 
```
rhdh:
  namespace: "" # the namespace of the RHDH instance (either an existing one or the one to be created by us)
  plugins:
    notificationsEmails: ""
```

Reply:
> Ultimately we can always remove this field from the CRD and make it dependent on having the CR installing RHDH or not. If the CR has rhdhOperator.enabled=false then the user is responsible of configuring the networkPolicy on their own (documentation would be needed to provide a sample manifest for the neworkPolicy).

We can be explicit about the need to specify both RHDH name and namespace.
```
rhdh:
  name: "" # the name of the RHDH instance (either an existing one or the one to be created by us)
  namespace: "" # the namespace of the RHDH instance (either an existing one or the one to be created by us)
```
then there isn't a need to guess the CR name.

Reply:
As agreed, moved the `existingRHDHNamespace` field to the new `networkPolicy` field and removed it  under the rhdhOperator. Still kept the more verbose name so the user knows it's specifically for existing RHDH instance.
The separation of the rhdh instance and operator will be moved to CRD proposal review and implemented in the feature PR and in the go-operator.

```
networkPolicy:
  existingRHDHNamespace: my-rhdh-operator
```